src/SAWScript/Bisimulation.hs

```diff
@@ -13,144 +13,561 @@ functionality in the future.
-greater of @N@ and @M@ steps have passed. The simpler relation type would not
-enable specification of this property.
-
-The main downside of this approach is that the resulting forall in the formula
-sent to the SMT solver quantifies over the initial output of the circuits prior
-to simulating a step. This can be confusing when reading the SAW source code,
-and could be resolved by requiring the user to provide two different relations
-(one over states, and one over states and outputs), but this puts more burden on
-the end user who would have to write two relations rather than just one. As
-such, we've chosen the approach that's easier on the end user, rather than the
-one that's easier on the SAW implementer.
+and considers @lhs@ and @rhs@ bisimilar when the following two theorems hold:
+* OUTPUT RELATION THEOREM:
+    forall s1 s2 in.
+      srel s1 s2 -> orel (lhs (s1, in)) (rhs (s2, in))
+* STATE RELATION THEOREM:
+    forall s1 s2 out1 out2.
+      orel (s1, out1) (s2, out2) -> srel s1 s2
+
+The OUTPUT RELATION THEOREM ensures that if the state relation holds prior to
+executing @lhs@ and @rhs@, then the output relation holds after executing @lhs@
+and @rhs@.  That is, the two terms step together.
+
+The STATE RELATION THEOREM ensures that if the output relation holds over some
+state and output value, then the state relation also holds over that same state.
+This ensures that the output relation captures the state relation.
+
+To enable compositionality, 'proveBisimulation' takes a list of proven
+bisimulation theorems as 'BisimTheorem's.  For each 'BisimTheorem', this
+implementation searches for invocations of the functions in the 'BisimTheorem'
+and replaces them with uninterpreted values that satisfy the 'BisimTheorem''s
+output relation.  Put another way, if @g_lhs@ contains some function @f_lhs@ and
+@g_rhs@ contains some function @f_rhs@ where @f_lhs@ and @f_rhs@ satisfy some
+output relation @f_orel@, then in the proof of some OUTPUT RELATION THEOREM for
+@g_lhs@ and @g_rhs@, the prover can replace @f_lhs@ with an uninterpreted value
+@v_lhs@ and @f_rhs@ with an uninterpreted value @v_rhs@.  Lastly, the prover
+assumes @v_lhs@ and @v_rhs@ satisfy @f_orel@.  In order to make this assumption,
+the state relation of the outer terms must ensure that the inner terms' state
+relation holds.  As such, the prover generates and proves a side condition for
+each applied 'BisimTheorem'.  Let @g_lhs_s@ and @g_rhs_s@ be the states for
+@g_lhs@ and @g_rhs@ respectively.  Additionally, let @f_lhs_s@ be the state for
+@f_lhs@ expressed as a substate of @g_rhs_s@, and @f_rhs_s@ be the state for
+@f_rhs@ expressed as a substate of @g_rhs_s@ (the prover examines the call sites
+for @f_lhs@ and @f_rhs@ within @g_lhs@ and @g_rhs@ to extract these @f@ states
+in terms of the @g@ states).  Lastly, let @g_srel@ be the state relation for the
+@g@ functions, and @f_srel@ be the state relation for the @f@ functions.  The
+prover then checks:
```

Comment:
This is a pretty long paragraph. Is there a way to split it up more?

Reply:
I broke this up a bit when I added some clarification around `f_lhs_s` and `f_rhs_s`